lib/rspec/mocks/exclude_stubbed_classes_from_subclasses.rb

```diff
@@ -0,0 +1,53 @@
+module RSpec
+  module Mocks
+    # Support for `exclude_stubbed_classes_from_subclasses` configuration.
+    #
+    # @private
+    class ExcludeStubbedClassesFromSubclasses
+      def self.enable!
+        return unless RUBY_VERSION >= "3.1"
+        return if Class.respond_to?(:subclasses_with_rspec_mocks)
+
+        require 'weakref'
+
+        mod_something = Module.new do
+          def excluded_subclasses
+            @excluded_subclasses ||= []
+            @excluded_subclasses.select(&:weakref_alive?).map do |ref|
+              begin
+                ref.__getobj__
+              rescue ::WeakRef::RefError
+                nil
+              end
+            end.compact
+          end
+
+          def exclude_subclass(constant)
+            @excluded_subclasses ||= []
+            @excluded_subclasses << ::WeakRef.new(constant)
+          end
+        end
+        RSpec::Mocks.extend(mod_something)
```

Comment:
Will we keep extending on every `enable!` (intermingled with disable!)?

Reply:
No, if we call `enable!` multiple times, it extends only once because of `return if Class.respond_to?(:subclasses_with_rspec_mocks)`.

Reply:
The problem I’m thinking of is when you enable!, disable!, and then enable! again when Class is not augmented, but the ExcludeStubbed… is.
Do you see what I mean?
You could add a described_class.disable! in between two enable! calls in the corresponding spec, and see how the `.once` expectation fails. I did not test this, just an assumption. I’ll be happy if it’s just my imagination.

What if we just define those two methods on the ExcludeStubbed immediately and don’t ‘extend’ it at all?

Reply:
ok, my bad. I take a look.

Reply:
I refactored the code. Doing and `extend` was now useless. I also changed and added missing specs.